### PR TITLE
Login: Add Content Security Policy header

### PR DESCRIPTION
**Pull request by yurynix**
### Background
[Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy) (CSP) serves as a mitigation against cross-site scripting attacks (XSS).

[More information in this blog post](https://hackernoon.com/im-harvesting-credit-card-numbers-and-passwords-from-your-site-here-s-how-9a8cb347c5b5)
Additional resources: [Google](https://developers.google.com/web/fundamentals/security/csp/) and [OWASP](https://www.owasp.org/index.php/Content_Security_Policy_Cheat_Sheet)

Tool to evaluate policy: https://csp-evaluator.withgoogle.com/

This PR aims to introduce the following policy to `/log-in/*`:
```
default-src 'self'; script-src 'self' 'unsafe-eval' stats.wp.com https://apis.google.com 'nonce-bfae467b9aaca24fc9acc9b0491128220dc551c30f487d6d6fd04523023270a83880c6583d7526593e3cf943fdce4fc2' 'nonce-b7c041b5271937931404f54ae3df321ac775ea8ea1d6ad00e1f109fdc2daaa4026cd2043d7a23cebaca1a35f06f73625' 'sha256-3yiQswl88knA3EhjrG5tj5gmV6EUdLYFvn2dygc0xUQ=' 'sha256-ZKTuGaoyrLu2lwYpcyzib+xE4/2mCN8PKv31uXS3Eg4='; base-uri 'none'; style-src 'self' *.wp.com https://fonts.googleapis.com; form-action 'self'; object-src 'none'; img-src 'self' *.wp.com https://www.google-analytics.com; frame-src 'self' https://public-api.wordpress.com https://accounts.google.com/; font-src 'self' *.wp.com https://fonts.gstatic.com data:; media-src 'self'; connect-src 'self' https://wordpress.com/; report-uri /cspreport
```
in a report only mode, for now.

While it would be beneficiary to have this for all of calypso, IMO we should start with something small and contained, such as the `/log-in/*` pages.

#### Testing instructions
  
1. Run `git checkout add/csp` and start your server, or open a [live branch](https://calypso.live/log-in?branch=add/csp)
2. Open the [`Login` page](http://calypso.localhost:3000/log-in)
3. Assert that you indeed see the `Content-Security-Policy-Report-Only` header
  
#### Reviews
  
- [ ] Code
- [ ] Product

### TODO ( in following PRs ):
- [ ] `store` library uses eval to get global object, can be solved by version upgrade, but for now it violates `'unsafe-eval'`
- [ ] `is-my-json-valid` depends on `generate-function` which violates `'unsafe-eval'` ( I don't see an obvious solution to this )
- [x]  Add hash/nonce to google analytics library to avoid whitelisting the host

### CODE REVIEW

**Comment by vortfu** [comment]
from https://csp-evaluator.withgoogle.com/
> [!] www.google-analytics.com is known to host JSONP endpoints which allow to bypass this CSP.

**Comment by yurynix** [comment]
> from https://csp-evaluator.withgoogle.com/
It's the 3rd item on the todo list =)

File: server/pages/index.js [hunk 1]
```diff
@@ -483,11 +544,7 @@ module.exports = function() {
 				} );
 
 				if ( ! section.isomorphic ) {
-					app.get(
-						pathRegex,
-						section.enableLoggedOut ? setUpRoute : setUpLoggedInRoute,
-						serverRender
-					);
+					app.get( pathRegex, setUpRoute, serverRender );
```
Comment (yurynix) [inline server/pages/index.js]: This allows moving `getDefaultContext()` to `setUpRoute()`,
we're anyway checking whether to call `setUpLoggedInRoute()` or `setUpLoggedOutRoute()`
in `setUpRoute()`

File: client/document/index.jsx [hunk 2]
```diff
@@ -45,8 +46,28 @@ class Document extends React.Component {
 			devDocs,
 			devDocsURL,
 			feedbackURL,
+			inlineScriptNonce,
+			analyticsScriptNonce,
 		} = this.props;
 
+		let inlineScript = `COMMIT_SHA = ${ jsonStringifyForHtml( commitSha ) };`;
```
Comment (Tug) [inline client/document/index.jsx]: could we add `\n` at the end of those? I think it would slightly be more readable when viewing the source.

File: server/pages/index.js [hunk 3]
```diff
@@ -498,6 +555,36 @@ module.exports = function() {
 			}
 		} );
 
+	app.post( '/cspreport', function( req, res ) {
```
Comment (Tug) [inline server/pages/index.js]: This could use some documentation, what is it used for?

File: server/pages/index.js [hunk 4]
```diff
@@ -342,13 +342,74 @@ function setUpLoggedInRoute( req, res, next ) {
...
+		'media-src': [ "'self'" ],
+		'connect-src': [ "'self'", 'https://wordpress.com/' ],
+		'report-uri': [ '/cspreport' ],
+	};
+
+	const policyString = Object.keys( policy )
+		.map( key => `${ key } ${ policy[ key ].join( ' ' ) }` )
+		.join( '; ' );
+
+	res.set( { 'Content-Security-Policy-Report-Only': policyString } );
```
Comment (Tug) [inline server/pages/index.js]: I see, we're in report only mode... We should still add some comments in the code to make that more obvious when reading the code

**Comment by Tug** [comment]
LGTM

There doesn't seem to be too much risk in running it in "report mode" so I say let's :shipit:

File: client/document/index.jsx [hunk 5]
```diff
@@ -45,8 +46,28 @@ class Document extends React.Component {
...
+			inlineScript += `var app = ${ jsonStringifyForHtml( app ) };\n`;
+		}
+
+		if ( initialReduxState ) {
+			inlineScript += `var initialReduxState = ${ jsonStringifyForHtml( initialReduxState ) };\n`;
+		}
+
+		if ( config ) {
+			inlineScript += config;
+		}
```
Comment (dmsnell) [inline client/document/index.jsx]: what's the purpose of these changes? they don't seem to be related to the CSP behaviors but maybe there's something obvious that isn't jumping out…

File: client/document/index.jsx [hunk 6]
```diff
@@ -198,6 +188,15 @@ class Document extends React.Component {
...
 					/>
+					{ // Load GA only if enabled in the config.
+					appConfig( 'google_analytics_enabled' ) && (
+						<script
+							async={ true }
+							type="text/javascript"
+							src="https://www.google-analytics.com/analytics.js"
+							nonce={ analyticsScriptNonce }
+						/>
+					) }
```
Comment (dmsnell) [inline client/document/index.jsx]: was there a merge in this PR that brought in unrelated changes?

File: server/pages/index.js [hunk 7]
```diff
@@ -342,13 +342,77 @@ function setUpLoggedInRoute( req, res, next ) {
...
-		setUpLoggedOutRoute( req, res, next );
+/**
+ * Sets up a Content Security Policy header
+ * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP
+ * @param {Object} req Express request object
+ * @param {Object} res Express response object
+ * @param {Function} next a callback to call when done
+ */
+function setUpCSP( req, res, next ) {
+	if ( ! req.originalUrl.match( /^\/log-in\// ) ) {
```
Comment (dmsnell) [inline server/pages/index.js]: if this is for security, maybe we should be more specific? does it matter if we are too broad in capturing the path?

```js
import URL form 'url'

const originalUrl = new URL( req.originalUrl );
if ( ! /^\/log-in\//.test( originalUrl.pathname ) ) {

}
```

File: client/document/index.jsx (hunk 5 above)
Comment (dmsnell) [reply client/document/index.jsx]: oh now I see it's to add the nonce, I think

how would you feel about also eliminating the mutation on the string. it's not necessary in any case

```js
const inlineScript = (
	`COMMIT_SHA = ${ … };\n` +
	( user ? `var currentUser = ${ … };\n` : '' ) +
	( app ? … : … ) +
	…
)
```

The thing I like about that is not having `inlineScript` ever in an invalid state and also not having to trace through the code to see if we might end up changing it lower down unexpectedly. As a `const` string we know it won't.

File: server/pages/index.js [hunk 8]
```diff
@@ -342,13 +342,77 @@ function setUpLoggedInRoute( req, res, next ) {
...
+		next();
+		return;
 	}
+
+	const inlineScripts = [
+		'sha256-3yiQswl88knA3EhjrG5tj5gmV6EUdLYFvn2dygc0xUQ=',
+		'sha256-ZKTuGaoyrLu2lwYpcyzib+xE4/2mCN8PKv31uXS3Eg4=',
+	];
```
Comment (dmsnell) [inline server/pages/index.js]: not sure where these are coming from, but it would be very very good to have a comment explaining from where and how to update them.

I was going to try and provide an example comment but actually I have no idea how we created them.

File: server/pages/index.js [hunk 9]
```diff
@@ -342,13 +342,77 @@ function setUpLoggedInRoute( req, res, next ) {
...
+		'report-uri': [ '/cspreport' ],
+	};
+
+	const policyString = Object.keys( policy )
+		.map( key => `${ key } ${ policy[ key ].join( ' ' ) }` )
+		.join( '; ' );
+
+	// For now we're just logging policy violations and not blocking them
+	// so we won't actually break anything, later we'll remove the 'Report-Only'
+	// part so browsers will block violating content.
```
Comment (dmsnell) [inline server/pages/index.js]: good comment!

File: server/pages/index.js [hunk 10]
```diff
@@ -342,13 +342,77 @@ function setUpLoggedInRoute( req, res, next ) {
...
+function setUpRoute( req, res, next ) {
+	req.context = getDefaultContext( req );
+	setUpCSP( req, res, () => {
+		if ( req.cookies.wordpress_logged_in ) {
+			// the user is probably logged in
+			setUpLoggedInRoute( req, res, next );
+		} else {
+			setUpLoggedOutRoute( req, res, next );
+		}
+	} );
```
Comment (dmsnell) [inline server/pages/index.js]: for fun:

```js
setUpCSP( req, res, () => req.cookies.wordpress_logged_in // a cookie probably indicates someone is logged-in
	? setUpLoggedInRoute( req, res, next )
	: setUpLoggedOutRoute( req, res, next )
)
```

File: server/pages/index.js [hunk 11]
```diff
@@ -342,13 +342,77 @@ function setUpLoggedInRoute( req, res, next ) {
...
-	} else {
-		setUpLoggedOutRoute( req, res, next );
+/**
+ * Sets up a Content Security Policy header
+ * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP
+ * @param {Object} req Express request object
+ * @param {Object} res Express response object
+ * @param {Function} next a callback to call when done
+ */
+function setUpCSP( req, res, next ) {
```
Comment (dmsnell) [inline server/pages/index.js]: nit pick you can ignore, but it `setUp` usually split at the `U`? I'm more accustomed to seeing `setupCSP` or even `setupCsp` but in any case `setup` as a single word.

File: server/pages/index.js [hunk 12]
```diff
@@ -498,6 +558,37 @@ module.exports = function() {
 			}
 		} );
 
+	// This used to log to tracks Content Security Policy violation reports sent by browsers
```
Comment (dmsnell) [inline server/pages/index.js]: Grammar note only for clarity here

"This used to" reads "At some point in the past we were using this function to log to Tracks"
"This is used to" read "We are using this function to log to Tracks"

Small word omission but a heavy impact on meaning.

File: server/pages/index.js [hunk 13]
```diff
@@ -498,6 +558,37 @@ module.exports = function() {
...
+				limit: '1mb',
+				encoding: 'utf-8',
+			},
+			function( err, body ) {
+				if ( err ) {
+					res.status( 500 ).send( 'Bad report!' );
+					return;
+				}
+
+				const jsonBody = JSON.parse( body );
```
Comment (dmsnell) [inline server/pages/index.js]: this is decidedly a _non_-JSON body, right? since w're parsing _out_ of a JSON string and _into_ a JavaScript object?

**Comment by dmsnell** [comment]
Thanks for working on this - a great boost to the project and not a task that can easily be done without focusing on it - bravo 👏

File: client/document/index.jsx (hunk 5 above)
Comment (yurynix) [reply client/document/index.jsx]: > oh now I see it's to add the nonce, I think

Indeed

> how would you feel about also eliminating the mutation on the string. it's not necessary in any case

Great idea, I'll update the PR

File: client/document/index.jsx (hunk 6 above)
Comment (yurynix) [reply client/document/index.jsx]: If you referring the above, that is done to add the `nonce` instead of whitelisting the domain `www.google-analytics.com`

File: server/pages/index.js (hunk 11 above)
Comment (yurynix) [reply server/pages/index.js]: I used the same case `setUpLoggedInRoute` and `setUpLoggedOutRoute` were already using

File: server/pages/index.js (hunk 7 above)
Comment (yurynix) [reply server/pages/index.js]: It's just so we will use that on `/log-in/*` only pages for now.

File: server/pages/index.js (hunk 8 above)
Comment (yurynix) [reply server/pages/index.js]: You're right, I'll add an explanation.

File: server/pages/index.js (hunk 13 above)
Comment (yurynix) [reply server/pages/index.js]: Not sure I understand your comment, you meant to say that `jsonBody` and `body` are actually inverse?
I'm changing that now. If you meant something else, do tell =)

Comment (dmsnell) [reply server/pages/index.js]: yeah that's what I meant. if I read a variable named `jsonBody` I would expect it to contain JSON data. in this case, it's' getting data that _was_ JSON but no longer is because it's parsed.

flipping the two matches the expectations I think much more closely.

File: server/pages/index.js (hunk 11 above)
Comment (dmsnell) [reply server/pages/index.js]: you prompted me to look it up :smile:

"setup" is a noun generally
"set up" is the verb form

for a reasonable explanation
https://www.visualthesaurus.com/cm/wc/is-it-set-up-set-up-or-setup/

in the several articles I read which seemed to harmonize quite well it
appears like your use of `setUpCSP()` is more appropriate for middleware
which _does_ things while it might take someone to the _`cspSetup`_ page,
for instance.

TIL

File: server/pages/index.js [hunk 14]
```diff
@@ -342,13 +342,84 @@ function setUpLoggedInRoute( req, res, next ) {
...
-		setUpLoggedOutRoute( req, res, next );
+/**
+ * Sets up a Content Security Policy header
+ * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP
+ * @param {Object} req Express request object
+ * @param {Object} res Express response object
+ * @param {Function} next a callback to call when done
+ */
+function setUpCSP( req, res, next ) {
+	const originalUrlPathname = req.originalUrl.split( '?' )[ 0 ];
```
Comment (dmsnell) [inline server/pages/index.js]: this is a reasonable compromise, though even more since we're on server code it might make sense to use `url` to actually parse it.

my primary concern was things that this takes care of, like `https://wordpress.com/some/path?redirect_to=/log-in/me` which I'm pretty sure wasn't the intended match of that original RegExp :+1:

Comment (yurynix) [reply server/pages/index.js]: Yeah, it was a good catch 👍

File: server/pages/index.js [hunk 15]
```diff
@@ -342,13 +342,84 @@ function setUpLoggedInRoute( req, res, next ) {
...
+
+	// We only setup CSP for /log-in* for now
+	if ( ! /^\/log-in/.test( originalUrlPathname ) ) {
+		next();
+		return;
 	}
+
+	// This is calculated by taking the contents of the script text from between the tags,
```
Comment (Tug) [inline server/pages/index.js]: I see, so could we generate it from `document/index.jsx` instead of using constants?

File: server/pages/index.js [hunk 16]
```diff
@@ -342,13 +342,84 @@ function setUpLoggedInRoute( req, res, next ) {
...
+ * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP
+ * @param {Object} req Express request object
+ * @param {Object} res Express response object
+ * @param {Function} next a callback to call when done
+ */
+function setUpCSP( req, res, next ) {
+	const originalUrlPathname = req.originalUrl.split( '?' )[ 0 ];
+
+	// We only setup CSP for /log-in* for now
+	if ( ! /^\/log-in/.test( originalUrlPathname ) ) {
```
Comment (Tug) [inline server/pages/index.js]: Just a note that by doing so, we're enabling it for `/log-in` but that does not mean it's going to run in all cases when accessing `/log-in` as this route can be loaded from the app without requesting the server.

**Comment by Tug** [comment]
I think it's ready for a v1.

@dmsnell what do you think?

**Comment by dmsnell** [comment]
@Tug sorry I don't have the cycles right now to review the updates. I'm comfortable with y'all doing what you think is best.
